---
title: "QAA Bi623"
author: "Bailey Knight"
date: "2023-09-11"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(tinytex)
library(kableExtra)

```
# Objective
The objectives of this assignment are to use existing tools for quality assessment and adaptor trimming, compare
the quality assessments to those from your own software, and to demonstrate your ability to summarize other
important information about this RNA-Seq data set in a high-level report. Questions for this assignment are answered in the discussion sections.\

## Background
We are given two paired end RNA seq reads from the 2017 bgmp RNAseq run. The files have already been demultiplexed.
the Data for all samples can be found in Talapas path "/projects/bgmp/shared/2017_sequencing/demultiplexed/"
The specific samples used for this report are: 8_2F_fox_S7_L008 and 14_3B_control_S10_L008.
Each has a Read 1 and Read 2 file associated with it. \

## General Workflow
The high level workflow is as follows: \
1. fastqc for quality reports \
2. Custom python script comparison to fastqc \
3. Trim adapters from reads using cutadapt \
4. Trim reads based on quality using trimmomatic \
5. Made script to plot read length distribution \
6. Make database using STAR software \
7. Align reads using STAR \
8. Count mapped and unmapped using python script \
9. Count genes mapped using htseq count for forward and reverse \
10. Count total genes mapped \

For a detailed account of the work see my lab notebook on Talapas below: \
/projects/bgmp/bailey/bioinfo/Bi623/Assignments/QAA/labnotebook.txt \
Ensemble fasta: https://ftp.ensembl.org/pub/release-110/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz
Ensemble GTG: https://ftp.ensembl.org/pub/release-110/gtf/mus_musculus/Mus_musculus.GRCm39.110.gtf.gz

# Part 1- Data Quality Investigation
```{r out.width="250px", echo=FALSE}
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/8_2F_fox_S7_L008_R1_001.fastq.gz_hist.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/8_2F_fox_S7_L008_R2_001.fastq.gz_hist.png")
```
**Figure 1:** Mean Quality Scores Per base position for each read of 8_2F_fox_S7_L008 Read 1 (left) and Read 2 (right). Plots were generated with custom python script.
```{r out.width="250px", echo=FALSE}
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/14_3B_control_S10_L008_R1_001.fastq.gz_hist.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/14_3B_control_S10_L008_R2_001.fastq.gz_hist.png")

```
**Figure 2:** Mean Quality Scores Per base position for each read of 14_3B_control_S10_L008 Read 1 (left) and Read 2 (right). Plots were generated with custom python script.
```{r out.width="250px", echo=FALSE}
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/8_2F_R1.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/8_2F_R1_Ns.png")

```

```{r out.width="250px", echo=FALSE}
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/8_2F_R2.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/8_2F_R2_Ns.png")
```
**Figure 3:** Left figures represent Quality Score Frequency per base position for sample 8_2F_fox_S7_L008 generated by Fastqc. Trend line follows the mean. Error bars represent data between 10th and 90th percentile. Box plots represent interquartile range. Right figures represent average N's found at each read position demonstrating a poor read at that position. Top figures represent Read 1 and bottom figures represent Read 2.
```{r out.width="250px", echo=FALSE}
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/14_3B_R1.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/14_3B_R1_Ns.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/14_3b_R2.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/14_3B_R2_Ns.png")
```
**Figure 4:** Left figures represent quality Score Frequency per base position for sample 14_3B_control_S10 generated by Fastqc. Trend line follows the mean. Error bars represent data between 10th and 90th percentile. Box plots represent interquartile range. Right figures represent average N's found at each read position demonstrating a poor read at that position. Top figures represent Read 1 and bottom figures represent Read 2. \


## Part 1 Discussion
14_3B plots shape look similar between the processing technique I coded vs fastqc. The plot type (mine being a histogram) vs fastqc line with error bars is different. The overall trends are pretty much identical with the max and mins occurring in the same spots on graphs. The max was around 40 for both processing techniques for the 14_3b data. The same conclusions can be drawn from the comparisons of the 8_2F data. In terms of processing time, fastqc was significantly faster and produced more plots. It took 3 minutes for the 8_2F_fox_S7 reads and 30 seconds for the 14_3B_control_S10 reads. Compared to my custom script which took 9 minutes and 3 minutes for the samples respectively.
Overall due to the quality scores generally being greater than 35 I would recommend this data to be used. Although read data is more variable this is expected considering the nature of the sequencing.  Proceed with greater caution with 14_3B_control as the per tile sequence quality had a streaked area for concern on the flow cell. Although this is a cause for concern, considering the overall quality score distribution the data can move forward for preprocessing and mapping. If data has poor results for mapping, it may be worth to filter out more poor quality cells.\

# Part 2 â€“ Adapter and Quality-based Trimming
```{r echo=FALSE}
cutadapt = data.frame(Sample =c('14_3B_control_s10','8_2F_fox_S7'),
                      Reads.Processed=c(4440378, 36482601),
                      Reads.Written=c(4440378, 36482601),
                      Percent.Read1.Trimmed=c(6.0,5.9),
                      Percent.Read2.Trimmed=c(6.7,6.6),
                      Percent.Written=c(100, 100))
                      
                      

kable(cutadapt, booktabs =T)%>%
  kable_styling(latex_options = c("hold_position", "scale_down"))
  
```
**Table 1:** Adapter Trimming using cutadapt for read pairs of both samples processed and written using cutadapt displayed in both read pairs and total basepairs. \

```{r, echo=FALSE}

trimmomatic = data.frame(Sample =c('14_3B_control_s10','8_2F_fox_S7'),
                      Iput.Read.Pairs=c(4440378, 36482601),
                      Surviving.Read.Pairs=c(4246652, 34791157),
                      Percent.Both.Surviving= c(95.64,95.36),
                      Percent.Only.Forward= c(4.11,4.47),
                      Percent.Only.Reverse= c(0.07,0.08),
                      Percent.Dropped= c(0.18,0.09))

kable(trimmomatic, booktabs =T)%>%
  kable_styling(latex_options = c("hold_position", "scale_down"))


```
**Table 2:** Quality Trimming using trimmomatic for read pairs of both samples. Pairs processed using trimmomatic displayed. Less than 1% reads dropped for both read pairs. \

```{r out.width="250px", echo=FALSE}
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/trimmed_R1_8_2F_fox_S7_L008_R1_001.fastq.gz_hist.png")
include_graphics("/Users/bailey/bioinfo/Bi623/Assignments/QAA/Figures/trimmed_R1_14_3B_control_S10_L008_R1_001.fastq.gz_hist.png")

```
**Figure 5:** Read Length distributions for two Paired End RNAseq samples, 8_2F_fox_S7 and 14_3B_control_s10. Samples were adapter trimmed with cutadapt. Sequences quality trimmed with Trimmomatic. \ 

## Part 2 Discussion
Cutadapt was used followed by Trimmomatic for adapter trimming and quality filtering. Setting for each can be found in labnotebook.txt referenced in "General Workflow" section of the report. Adapter trimming trimmed 6% and 5.9% of sequences for reads 1 and 2 respectively for 14_3B_control_s10. Adapter trimming trimmed 6.7% and 6.6% of sequences for reads 1 and 2 respectively for 8_2F_fox_S7. Trimmomatic quality filtered out a total of 0.09% and 0.18% reads for 8_2F_fox_S7 and 14_3B_control_s10 resepctively. The adapter trimming for both reads was nearly identical for the forward and reverse reads as expected since they are compliments. The quality forward reads were trimmed at a higher rate than the reverse reads by nearly two orders of magnitude. Read 2 had consistently greater frequency of low length reads than Read 1 as expected considering idle time for read 2 on the sequencer while read 1 is being read. \

# Part 3 Alignment and Mapping
```{r echo=FALSE}
Percent_mapped14 = (8312388/(180916+8312388)*100)
Percent_mapped8 = (67070899/(2511415+67070899)*100)
Mapped_reads = data.frame(Sample =c('14_3B_control_s10','8_2F_fox_S7'),
                            Mapped = c(8312388, 67070899 ),
                            UnMapped= c(180916, 2511415),
                            Percent.Mapped = c(Percent_mapped14, Percent_mapped8))

kable(Mapped_reads, booktabs =T)%>%
  kable_styling(font_size = 10, latex_options = c("hold_position", "striped"))

```
**Table3:** Percent mapped reads for
```{r echo=FALSE}
Percent_Mapped = data.frame(Sample =c('14_3B_control_s10','8_2F_fox_S7'),
                            Percent.Mapped.htseq.Reverse = c(86.3475, 80.5948),
                            Percent.Mapped.htseq.Yes= c(3.87395, 3.624))

kable(Percent_Mapped, booktabs =T)%>%
  kable_styling(font_size = 10, latex_options = c("hold_position", "striped"))

```
**Table 4:** Post-mapping analysis of htseq parameters for read counts that map to features. Stranded parameters used were "--stranded=yes" and "--stranded=reverse". Percents represent percent of reads that mapped to featurs using the parameters previously specified.
Command used to generate percentages: awk '{total+=$2} $1~"__" {sum+=$2} END {print (total-sum)/total*100}' filepath.

## Part 3 Discussion
Mouse Genome fasta and GTF files were acquired from ensemble. Links can be found under the "General Workflow" section. A database was generated using STAR software and the two files mentioned previously. The trimmed and filtered sequences of each sample were then aligned against the database using the STAR software. A SAM file was generated from the alignment that was then run through a custom python script. The script counted mapped and unmapped reads while ignoring any seconday alignments. Htseq was run on the SAM files using "--stranded=yes" and "--stranded=reverse" and mapping percentages were calculated. "stranded--reverse" yielded 80.6% and 86.9% mapped compared to 3.6% and 3.9% mapped for 8_2Freverse and 14_3b_control respectively. This indicates the RNAseq run is paired end and stranded run because the "reverse" option takes read order and orientation into account during comparison.
